### PR TITLE
SP-4346 Resource validation

### PR DESCRIPTION
**Pull request by Idris0v**

### CODE REVIEW

File: Provider Provisioner/src/handle-input-offer.ts [hunk 1]
```diff
@@ -28,19 +35,49 @@ export async function handleInputOffer(params: HandleInputOfferParams): Promise<
   if (!orders.length) {
     return;
   }
+  await spctlService.initialize();
```
Comment (DmitrySmv) [inline Provider Provisioner/src/handle-input-offer.ts:38]: It looks weird. The `spctlService` must be initialized before it is passed to the `handleInputOffer` function.

File: Provider Provisioner/src/handle-input-offer.ts [hunk 2]
```diff
@@ -28,19 +35,49 @@ export async function handleInputOffer(params: HandleInputOfferParams): Promise<
...
 
   const resourceJsonPath = 'resource.json';
   await fs.writeFile(resourceJsonPath, JSON.stringify(inputOffer.resourceFileContent), 'utf-8');
 
+  await Promise.allSettled(
```
Comment (DmitrySmv) [inline Provider Provisioner/src/handle-input-offer.ts:43]:
```suggestion
  await Promise.all(
```
I noticed that you’re already catching all errors in your promises. In such cases, the use of `.allSettled` might not be necessary.

File: Provider Provisioner/src/handle-input-offer.ts [hunk 3]
```diff
@@ -28,19 +35,49 @@ export async function handleInputOffer(params: HandleInputOfferParams): Promise<
...
+        if (err instanceof StorageResourceValidationError) {
+          log.error({ err }, `Completing order ${order.id} with status Error`);
+          return await completeWithError({
+            orderId: order.id,
+            errorMessage: err.message,
+            spctlService,
+          }).catch((err) => log.error({ err }, 'Completing order with Error failed. Ignoring...'));
+        }
+
+        log.error({ err }, `Failed to complete order ${order.id}. Ignoring..`);
```
Comment (DmitrySmv) [inline Provider Provisioner/src/handle-input-offer.ts:61]:
```suggestion
        log.error({ err }, `Failed to complete order ${order.id} with status Done`);
```
Please move this log after 51 line.

**Review by DmitrySmv — CHANGES_REQUESTED** [review]

File: Provider Provisioner/src/spctl/types.ts [hunk 4]
```diff
@@ -23,7 +23,7 @@ export interface GetOrdersParams {
 }
 
 export interface CompleteOrdersParams {
```
Comment (DmitrySmv) [inline Provider Provisioner/src/spctl/types.ts:25]:
```suggestion
export interface CompleteOrderParams {
```

File: Provider Provisioner/src/spctl/spctl.service.ts [hunk 5]
```diff
@@ -41,28 +64,26 @@ export class SpctlService {
     return savedResult.list;
   }
 
-  async completeOrders(params: CompleteOrdersParams): Promise<void> {
+  async completeOrder(params: CompleteOrdersParams): Promise<void> {
```
Comment (DmitrySmv) [inline Provider Provisioner/src/spctl/spctl.service.ts:67]:
```suggestion
  async completeOrder(params: CompleteOrderParams): Promise<void> {
```

File: Provider Provisioner/src/handle-input-offer.ts [hunk 6]
```diff
@@ -28,19 +35,49 @@ export async function handleInputOffer(params: HandleInputOfferParams): Promise<
...
+      }
+    }),
+  );
+
+  await fs.rm(resourceJsonPath, { force: true });
+}
+
+async function completeWithError(params: CompleteOrderWithErrorParams) {
```
Comment (DmitrySmv) [inline Provider Provisioner/src/handle-input-offer.ts:69]:
```suggestion
async function completeOrderWithError(params: CompleteOrderWithErrorParams): Promise<void> {
```

File: Provider Provisioner/src/handle-input-offer.ts (hunk 1 above)
Comment (Idris0v) [reply Provider Provisioner/src/handle-input-offer.ts:38]: We're initializing blockchain here. we dont need to do it before we actually have some orders to complete. Will rename it to initializeBlockchainConnector

File: Provider Provisioner/src/spctl/spctl.service.ts [hunk 7]
```diff
@@ -1,19 +1,42 @@
...
+    const blockchainConfig = {
+      contractAddress: config.blockchain.smartContractAddress,
+      blockchainUrl: config.blockchain.rpcUrl,
+    };
+
+    this.actionAccountAddress = await initBlockchainConnector({
+      blockchainConfig,
+      actionAccountKey: config.blockchain.accountPrivateKey,
+    });
+    this.initialized = true;
```
Comment (DmitrySmv) [inline Provider Provisioner/src/spctl/spctl.service.ts:37]:
```suggestion
    this.blockchainConnectorInited = true;
```

File: Provider Provisioner/src/spctl/spctl.service.ts [hunk 8]
```diff
@@ -41,28 +64,26 @@ export class SpctlService {
     return savedResult.list;
   }
 
-  async completeOrders(params: CompleteOrdersParams): Promise<void> {
+  async completeOrder(params: CompleteOrderParams): Promise<void> {
+    if (!this.initialized) {
+      throw new Error('Spctl service is not initialized');
```
Comment (DmitrySmv) [inline Provider Provisioner/src/spctl/spctl.service.ts:69]:
```suggestion
    if (!this.blockchainConnectorInited) {
      throw new Error('Blockchain connector is not inited');
```

**Review by DmitrySmv — CHANGES_REQUESTED** [review]

**Review by DmitrySmv — APPROVED** [review]

**Pull request merged**